File: apps/frontend/src/main/colored-logs.ts
/**
 * Colored Logs Utility for Frontend
 * =================================
 * 
 * Utility for differentiating frontend logs with distinct colors.
 * Provides purple/pink theme to distinguish from backend blue/cyan theme.
 * 
 * Usage:
 * import { frontendLog } from './colored-logs';
 * 
 * frontendLog.debug('Main process started');
 * frontendLog.info('Window created');
 * frontendLog.error('Failed to load config');
 */

import { app } from 'electron';

// Helper function to get model info without circular dependency
function getModelInfoForLogs(): { provider: string; modelLabel: string } {
  try {
    // Try to get settings directly to avoid circular import
    const { readSettingsFile } = require('./settings-utils');
    const { PROVIDER_MODELS_MAP } = require('../shared/constants/models');
    
    const settings = readSettingsFile();
    
    if (settings && settings.selectedProvider && settings.defaultModel) {
      const provider = settings.selectedProvider as string;
      const modelId = settings.defaultModel as string;
      
      // Try to get model label from PROVIDER_MODELS_MAP
      const providerModels = PROVIDER_MODELS_MAP[provider];
      let modelLabel = modelId;
      
      if (providerModels) {
        const modelInfo = providerModels.find((m: any) => m.value === modelId);
        if (modelInfo) {
          modelLabel = modelInfo.label;
        }
      }
      
      return { provider, modelLabel };
    }
    
    // Fallback: try to detect from environment like backend does
    return detectProviderFromEnv();
    
  } catch (error) {
    // Final fallback: try environment detection
    try {
      return detectProviderFromEnv();
    } catch {
      return { provider: 'error', modelLabel: 'error' };
    }
  }
}

// Helper function to detect provider from environment (backend-style fallback)
function detectProviderFromEnv(): { provider: string; modelLabel: string } {
  const env = process.env;
  
  if (env.ANTHROPIC_API_KEY || env.CLAUDE_API_KEY || env.CLAUDE_CODE_OAUTH_TOKEN) {
    return { provider: 'anthropic', modelLabel: 'Claude' };
  } else if (env.OPENAI_API_KEY) {
    return { provider: 'openai', modelLabel: 'ChatGPT' };
  } else if (env.GOOGLE_API_KEY) {
    return { provider: 'google', modelLabel: 'Gemini' };
  } else if (env.GROK_API_KEY) {
    return { provider: 'grok', modelLabel: 'Grok' };
  } else if (env.OLLAMA_BASE_URL) {
    return { provider: 'ollama', modelLabel: 'Ollama' };
  } else if (env.WINDSURF_API_KEY || env.WINDSURF_OAUTH_TOKEN || env.CODEIUM_API_KEY) {
    return { provider: 'windsurf', modelLabel: 'Windsurf (Codeium)' };
  } else if (env.SELECTED_LLM_PROVIDER) {
    // Catch-all: honor the explicit provider selection from UI
    const p = env.SELECTED_LLM_PROVIDER.toLowerCase();
    return { provider: p, modelLabel: p.charAt(0).toUpperCase() + p.slice(1) };
  } else {
    // Default to copilot since we're in Claude Code environment
    return { provider: 'copilot', modelLabel: 'GitHub Copilot' };
  }
}

// ANSI color codes for terminal output
class FrontendColors {
  static readonly RESET = '\x1b[0m';
  static readonly BOLD = '\x1b[1m';
  static readonly DIM = '\x1b[2m';
  
  // Frontend color scheme - Purple/Pink theme
  static readonly PREFIX = '\x1b[95m';    // Magenta
  static readonly DEBUG = '\x1b[95m';     // Magenta
  static readonly DEBUG_DIM = '\x1b[38;5;183m';  // Light magenta
  static readonly INFO = '\x1b[38;5;147m';      // Light purple
  static readonly SUCCESS = '\x1b[38;5;120m';   // Teal green
  static readonly WARNING = '\x1b[38;5;221m';   // Light yellow
  static readonly ERROR = '\x1b[38;5;196m';     // Red
  
  // Common colors
  static readonly TIMESTAMP = '\x1b[90m';  // Gray
  static readonly MODULE = '\x1b[33m';     // Yellow  
  static readonly KEY = '\x1b[35m';        // Magenta
  static readonly VALUE = '\x1b[37m';      // White
}

function supportsColor(): boolean {
  // Check for common environment variables
  if (process.env.NO_COLOR) {
    return false;
  }

  if (process.env.FORCE_COLOR) {
    return true;
  }
  
  // Check if we're in a terminal
  if (!process.stdout.isTTY) {
    return false;
  }
  
  // Check TERM environment variable
  const term = process.env.TERM || '';
  if (term === 'dumb' || term === 'unknown') {
    return false;
  }
  
  // Check for common color-supporting terminals
  return /color|256|xterm|screen|tmux/i.test(term);
}

function getTimestamp(): string {
  return new Date().toLocaleTimeString('en-US', { 
    hour12: false,
    hour: '2-digit',
    minute: '2-digit',
    second: '2-digit',
    fractionalSecondDigits: 3
  });
}

function formatValue(value: any, maxLength: number = 200): string {
  if (value === null || value === undefined) {
    return 'None';
  }
  
  if (typeof value === 'object') {
    try {
      const formatted = JSON.stringify(value, null, 2);
      return formatted.length > maxLength ? formatted.substring(0, maxLength) + '...' : formatted;
    } catch (error) {
      return String(value).substring(0, maxLength);
    }
  }
  
  const strValue = String(value);
  return strValue.length > maxLength ? strValue.substring(0, maxLength) + '...' : strValue;
}

function formatFrontendLog(
  message: string,
  level: string = 'DEBUG',
  module: string = 'frontend',
  timestamp?: string,
  ...kwargs: any[]
): string {
  if (!supportsColor()) {
    // Return non-colored version for non-color terminals
    const modelInfo = getModelInfoForLogs();
    const modelInfoString = `[${modelInfo.provider}:${modelInfo.modelLabel}]`;
    return `[${timestamp || getTimestamp()}] [FRONTEND] [${level}] [${module}] ${modelInfoString} ${message}`;
  }
  
  const ts = timestamp || getTimestamp();
  const modelInfo = getModelInfoForLogs();
  const modelInfoString = `${FrontendColors.RESET}[${modelInfo.provider}:${modelInfo.modelLabel}]${FrontendColors.RESET}`;
  
  // Get the appropriate color for the level
  const levelColors: Record<string, string> = {
    'DEBUG': FrontendColors.DEBUG,
    'INFO': FrontendColors.INFO,
    'SUCCESS': FrontendColors.SUCCESS,
    'WARNING': FrontendColors.WARNING,
    'ERROR': FrontendColors.ERROR,
  };
  const levelColor = levelColors[level.toUpperCase()] || FrontendColors.DEBUG;
  
  // Build the log line with model info
  const parts = [
    `${FrontendColors.TIMESTAMP}[${ts}]${FrontendColors.RESET}`,
    `${FrontendColors.PREFIX}[FRONTEND]${FrontendColors.RESET}`,
    `${levelColor}[${level}]${FrontendColors.RESET}`,
    `${FrontendColors.MODULE}[${module}]${FrontendColors.RESET}`,
    modelInfoString,
    `${FrontendColors.DEBUG_DIM}${message}${FrontendColors.RESET}`,
  ];
  
  let logLine = parts.join(' ');
  
  // Add additional arguments on separate lines if present
  if (kwargs.length > 0) {
    kwargs.forEach((arg, index) => {
      const formattedValue = formatValue(arg);
      if (formattedValue.includes('\n')) {
        // Multi-line value
        logLine += `\n  ${FrontendColors.KEY}arg${index}${FrontendColors.RESET}:`;
        formattedValue.split('\n').forEach(line => {
          logLine += `\n    ${FrontendColors.VALUE}${line}${FrontendColors.RESET}`;
        });
      } else {
        logLine += `\n  ${FrontendColors.KEY}arg${index}${FrontendColors.RESET}: ${FrontendColors.VALUE}${formattedValue}${FrontendColors.RESET}`;
      }
    });
  }
  
  return logLine;
}

function writeFrontendLog(message: string, level: string = 'DEBUG', module?: string, ...kwargs: any[]): void {
  const formatted = formatFrontendLog(message, level, module, undefined, ...kwargs);
  // Use process.stderr.write to avoid recursion with overridden console methods
  if (process.stderr && process.stderr.write) {
    process.stderr.write(formatted + '\n');
  } else {
    // Fallback to original console if available
    const originalConsole = (console as any).original;
    if (originalConsole && originalConsole.error) {
      originalConsole.error(formatted);
    } else {
      // Last resort - use writeFileSync
      try {
        require('fs').writeFileSync(1, formatted + '\n');
      } catch (e) {
        // If all else fails, stay silent to avoid infinite recursion
      }
    }
  }
}

// Main frontend logger object
export const frontendLog = {
  debug: (message: string, ...kwargs: any[]) => {
    writeFrontendLog(message, 'DEBUG', 'frontend', ...kwargs);
  },
  
  info: (message: string, ...kwargs: any[]) => {
    writeFrontendLog(message, 'INFO', 'frontend', ...kwargs);
  },
  
  success: (message: string, ...kwargs: any[]) => {
    writeFrontendLog(message, 'SUCCESS', 'frontend', ...kwargs);
  },
  
  warn: (message: string, ...kwargs: any[]) => {
    writeFrontendLog(message, 'WARNING', 'frontend', ...kwargs);
  },
  
  error: (message: string, ...kwargs: any[]) => {
    writeFrontendLog(message, 'ERROR', 'frontend', ...kwargs);
  },
  
  // Module-specific loggers
  main: {
    debug: (message: string, ...kwargs: any[]) => {
      writeFrontendLog(message, 'DEBUG', 'main', ...kwargs);
    },
    info: (message: string, ...kwargs: any[]) => {
      writeFrontendLog(message, 'INFO', 'main', ...kwargs);
    },
    success: (message: string, ...kwargs: any[]) => {
      writeFrontendLog(message, 'SUCCESS', 'main', ...kwargs);
    },
    warn: (message: string, ...kwargs: any[]) => {
      writeFrontendLog(message, 'WARNING', 'main', ...kwargs);
    },
    error: (message: string, ...kwargs: any[]) => {
      writeFrontendLog(message, 'ERROR', 'main', ...kwargs);
    },
  },
  
  renderer: {
    debug: (message: string, ...kwargs: any[]) => {
      writeFrontendLog(message, 'DEBUG', 'renderer', ...kwargs);
    },
    info: (message: string, ...kwargs: any[]) => {
      writeFrontendLog(message, 'INFO', 'renderer', ...kwargs);
    },
    success: (message: string, ...kwargs: any[]) => {
      writeFrontendLog(message, 'SUCCESS', 'renderer', ...kwargs);
    },
    warn: (message: string, ...kwargs: any[]) => {
      writeFrontendLog(message, 'WARNING', 'renderer', ...kwargs);
    },
    error: (message: string, ...kwargs: any[]) => {
      writeFrontendLog(message, 'ERROR', 'renderer', ...kwargs);
    },
  },
  
  ipc: {
    debug: (message: string, ...kwargs: any[]) => {
      writeFrontendLog(message, 'DEBUG', 'ipc', ...kwargs);
    },
    info: (message: string, ...kwargs: any[]) => {
      writeFrontendLog(message, 'INFO', 'ipc', ...kwargs);
    },
    success: (message: string, ...kwargs: any[]) => {
      writeFrontendLog(message, 'SUCCESS', 'ipc', ...kwargs);
    },
    warn: (message: string, ...kwargs: any[]) => {
      writeFrontendLog(message, 'WARNING', 'ipc', ...kwargs);
    },
    error: (message: string, ...kwargs: any[]) => {
      writeFrontendLog(message, 'ERROR', 'ipc', ...kwargs);
    },
  },
};

// Export convenience functions
export const logFrontendDebug = (message: string, ...kwargs: any[]) => frontendLog.debug(message, ...kwargs);
export const logFrontendInfo = (message: string, ...kwargs: any[]) => frontendLog.info(message, ...kwargs);
export const logFrontendSuccess = (message: string, ...kwargs: any[]) => frontendLog.success(message, ...kwargs);
export const logFrontendWarning = (message: string, ...kwargs: any[]) => frontendLog.warn(message, ...kwargs);
export const logFrontendError = (message: string, ...kwargs: any[]) => frontendLog.error(message, ...kwargs);

// NOTE: Console override disabled to prevent recursion issues
// Use frontendLog directly instead of overriding console methods
/*
// Enhanced console object override for development
if (process.env.NODE_ENV === 'development') {
  const originalConsole = {
    log: console.log,
    debug: console.debug,
    info: console.info,
    warn: console.warn,
    error: console.error,
  };
  
  // Override console methods to use colored logging
  console.log = (...args: any[]) => {
    if (args.length > 0) {
      frontendLog.info(String(args[0]), ...args.slice(1));
    } else {
      originalConsole.log(...args);
    }
  };
  
  console.debug = (...args: any[]) => {
    if (args.length > 0) {
      frontendLog.debug(String(args[0]), ...args.slice(1));
    } else {
      originalConsole.debug(...args);
    }
  };
  
  console.info = (...args: any[]) => {
    if (args.length > 0) {
      frontendLog.info(String(args[0]), ...args.slice(1));
    } else {
      originalConsole.info(...args);
    }
  };
  
  console.warn = (...args: any[]) => {
    if (args.length > 0) {
      frontendLog.warn(String(args[0]), ...args.slice(1));
    } else {
      originalConsole.warn(...args);
    }
  };
  
  console.error = (...args: any[]) => {
    if (args.length > 0) {
      frontendLog.error(String(args[0]), ...args.slice(1));
    } else {
      originalConsole.error(...args);
    }
  };
  
  // Keep original methods available if needed
  (console as any).original = originalConsole;
}
*/

export default frontendLog;

File: apps/backend/core/colored_logs.py
#!/usr/bin/env python3
"""
Colored Logs Utility
====================

Utility for differentiating backend and frontend logs in the console.
Provides distinct color schemes for each component to avoid log confusion.

Usage:
    from colored_logs import get_backend_colors, get_frontend_colors
    
    backend_colors = get_backend_colors()
    frontend_colors = get_frontend_colors()
    
    # Backend logs
    print(f"{backend_colors.DEBUG}[BACKEND DEBUG]{backend_colors.RESET} Message")
    
    # Frontend logs  
    print(f"{frontend_colors.DEBUG}[FRONTEND DEBUG]{frontend_colors.RESET} Message")
"""

import os
import sys
from typing import Dict


class LogColors:
    """Base class for log color schemes."""
    
    RESET = "\033[0m"
    BOLD = "\033[1m"
    DIM = "\033[2m"
    
    # Common colors
    TIMESTAMP = "\033[90m"  # Gray
    MODULE = "\033[33m"     # Yellow  
    KEY = "\033[35m"        # Magenta
    VALUE = "\033[37m"     # White
    
    # Default level colors (can be overridden by subclasses)
    DEBUG = "\033[36m"      # Cyan
    INFO = "\033[94m"       # Bright blue
    SUCCESS = "\033[92m"   # Green
    WARNING = "\033[93m"   # Yellow
    ERROR = "\033[91m"     # Red
    
    # Prefix for component identification
    PREFIX = "\033[94m"     # Bright blue (default)
    DEBUG_DIM = "\033[96m"  # Light cyan


class BackendColors(LogColors):
    """Color scheme for backend logs - Blue/Cyan theme."""
    
    PREFIX = "\033[94m"    # Bright blue
    DEBUG = "\033[36m"     # Cyan
    DEBUG_DIM = "\033[96m" # Light cyan
    INFO = "\033[94m"      # Bright blue
    SUCCESS = "\033[92m"   # Green
    WARNING = "\033[93m"   # Yellow
    ERROR = "\033[91m"     # Red


class FrontendColors(LogColors):
    """Color scheme for frontend logs - Purple/Pink theme."""
    
    PREFIX = "\033[95m"    # Magenta
    DEBUG = "\033[95m"     # Magenta
    DEBUG_DIM = "\033[38;5;183m"  # Light magenta
    INFO = "\033[38;5;147m"  # Light purple
    SUCCESS = "\033[38;5;120m"  # Teal green
    WARNING = "\033[38;5;221m"  # Light yellow
    ERROR = "\033[38;5;196m"    # Red


def _supports_color() -> bool:
    """Check if the terminal supports ANSI colors."""
    # Check for common environment variables
    if os.environ.get("NO_COLOR"):
        return False
    if os.environ.get("FORCE_COLOR"):
        return True
    
    # Check if we're in a terminal
    if not hasattr(sys.stdout, "isatty") or not sys.stdout.isatty():
        return False
    
    # Check TERM environment variable
    term = os.environ.get("TERM", "")
    if term in ("dumb", "unknown", "unknown"):
        return False
    
    # Check for common color-supporting terminals
    if any(color_term in term for color_term in ["color", "256", "xterm", "screen", "tmux"]):
        return True
    
    return False


def get_backend_colors() -> LogColors:
    """Get the backend color scheme."""
    if _supports_color():
        return BackendColors()
    else:
        # Return a colorless version for non-color terminals
        return LogColors()


def get_frontend_colors() -> LogColors:
    """Get the frontend color scheme."""
    if _supports_color():
        return FrontendColors()
    else:
        # Return a colorless version for non-color terminals
        return LogColors()


def format_backend_log(
    message: str,
    level: str = "DEBUG",
    module: str = "backend",
    timestamp: str = None,
    **kwargs
) -> str:
    """
    Format a backend log message with appropriate colors and model info.
    
    Args:
        message: The log message
        level: Log level (DEBUG, INFO, SUCCESS, WARNING, ERROR)
        module: Source module name
        timestamp: Optional timestamp (will generate if not provided)
        **kwargs: Additional key-value pairs to include
    
    Returns:
        Formatted colored log message
    """
    colors = get_backend_colors()
    
    if timestamp is None:
        from datetime import datetime
        timestamp = datetime.now().strftime("%H:%M:%S.%f")[:-3]
    
    # Get model information for logging
    try:
        from .model_info import get_model_info_for_logs
        model_info = get_model_info_for_logs()
        model_info_string = f"{colors.RESET}[{model_info['provider']}:{model_info['model_label']}]{colors.RESET}"
    except Exception:
        # Fallback if model info is not available
        model_info_string = f"{colors.RESET}[unknown:unknown]{colors.RESET}"
    
    # Get the appropriate color for the level
    level_colors = {
        "DEBUG": colors.DEBUG,
        "INFO": colors.INFO,
        "SUCCESS": colors.SUCCESS,
        "WARNING": colors.WARNING,
        "ERROR": colors.ERROR,
    }
    level_color = level_colors.get(level.upper(), colors.DEBUG)
    
    # Build the log line
    parts = [
        f"{colors.TIMESTAMP}[{timestamp}]{colors.RESET}",
        f"{colors.PREFIX}[BACKEND]{colors.RESET}",
        f"{level_color}[{level}]{colors.RESET}",
        f"{colors.MODULE}[{module}]{colors.RESET}",
        model_info_string,
        f"{colors.DEBUG_DIM}{message}{colors.RESET}",
    ]
    
    log_line = " ".join(parts)
    
    # Add kwargs on separate lines if present
    if kwargs:
        for key, value in kwargs.items():
            formatted_value = _format_value(value)
            if "\n" in formatted_value:
                # Multi-line value
                log_line += f"\n  {colors.KEY}{key}{colors.RESET}:"
                for line in formatted_value.split("\n"):
                    log_line += f"\n    {colors.VALUE}{line}{colors.RESET}"
            else:
                log_line += f"\n  {colors.KEY}{key}{colors.RESET}: {colors.VALUE}{formatted_value}{colors.RESET}"
    
    return log_line


def format_frontend_log(
    message: str,
    level: str = "DEBUG",
    module: str = "frontend",
    timestamp: str = None,
    **kwargs
) -> str:
    """
    Format a frontend log message with appropriate colors.
    
    Args:
        message: The log message
        level: Log level (DEBUG, INFO, SUCCESS, WARNING, ERROR)
        module: Source module name
        timestamp: Optional timestamp (will generate if not provided)
        **kwargs: Additional key-value pairs to include
    
    Returns:
        Formatted colored log message
    """
    colors = get_frontend_colors()
    
    if timestamp is None:
        from datetime import datetime
        timestamp = datetime.now().strftime("%H:%M:%S.%f")[:-3]
    
    # Get the appropriate color for the level
    level_colors = {
        "DEBUG": colors.DEBUG,
        "INFO": colors.INFO,
        "SUCCESS": colors.SUCCESS,
        "WARNING": colors.WARNING,
        "ERROR": colors.ERROR,
    }
    level_color = level_colors.get(level.upper(), colors.DEBUG)
    
    # Build the log line
    parts = [
        f"{colors.TIMESTAMP}[{timestamp}]{colors.RESET}",
        f"{colors.PREFIX}[FRONTEND]{colors.RESET}",
        f"{level_color}[{level}]{colors.RESET}",
        f"{colors.MODULE}[{module}]{colors.RESET}",
        f"{colors.DEBUG_DIM}{message}{colors.RESET}",
    ]
    
    log_line = " ".join(parts)
    
    # Add kwargs on separate lines if present
    if kwargs:
        for key, value in kwargs.items():
            formatted_value = _format_value(value)
            if "\n" in formatted_value:
                # Multi-line value
                log_line += f"\n  {colors.KEY}{key}{colors.RESET}:"
                for line in formatted_value.split("\n"):
                    log_line += f"\n    {colors.VALUE}{line}{colors.RESET}"
            else:
                log_line += f"\n  {colors.KEY}{key}{colors.RESET}: {colors.VALUE}{formatted_value}{colors.RESET}"
    
    return log_line


def _format_value(value, max_length: int = 200) -> str:
    """Format a value for debug output, truncating if necessary."""
    import json
    
    if value is None:
        return "None"
    
    if isinstance(value, (dict, list)):
        try:
            formatted = json.dumps(value, indent=2, default=str)
            if len(formatted) > max_length:
                formatted = formatted[:max_length] + "..."
            return formatted
        except (TypeError, ValueError):
            return str(value)[:max_length]
    
    str_value = str(value)
    if len(str_value) > max_length:
        return str_value[:max_length] + "..."
    return str_value


def write_backend_log(message: str, level: str = "DEBUG", **kwargs) -> None:
    """Write a backend log message to stderr."""
    formatted = format_backend_log(message, level, **kwargs)
    print(formatted, file=sys.stderr)


def write_frontend_log(message: str, level: str = "DEBUG", **kwargs) -> None:
    """Write a frontend log message to stderr."""
    formatted = format_frontend_log(message, level, **kwargs)
    print(formatted, file=sys.stderr)


# Convenience functions
def backend_debug(message: str, **kwargs) -> None:
    """Log a backend debug message."""
    write_backend_log(message, "DEBUG", **kwargs)


def backend_info(message: str, **kwargs) -> None:
    """Log a backend info message."""
    write_backend_log(message, "INFO", **kwargs)


def backend_success(message: str, **kwargs) -> None:
    """Log a backend success message."""
    write_backend_log(message, "SUCCESS", **kwargs)


def backend_warning(message: str, **kwargs) -> None:
    """Log a backend warning message."""
    write_backend_log(message, "WARNING", **kwargs)


def backend_error(message: str, **kwargs) -> None:
    """Log a backend error message."""
    write_backend_log(message, "ERROR", **kwargs)


def frontend_debug(message: str, **kwargs) -> None:
    """Log a frontend debug message."""
    write_frontend_log(message, "DEBUG", **kwargs)


def frontend_info(message: str, **kwargs) -> None:
    """Log a frontend info message."""
    write_frontend_log(message, "INFO", **kwargs)


def frontend_success(message: str, **kwargs) -> None:
    """Log a frontend success message."""
    write_frontend_log(message, "SUCCESS", **kwargs)


def frontend_warning(message: str, **kwargs) -> None:
    """Log a frontend warning message."""
    write_frontend_log(message, "WARNING", **kwargs)


def frontend_error(message: str, **kwargs) -> None:
    """Log a frontend error message."""
    write_frontend_log(message, "ERROR", **kwargs)
